---
title: "CAEN Tutorial"
author: "Yan Zhou"
package: CAEN
date: "`r Sys.Date()`"
output: rmarkdown::html_document
vignette: >
    %\VignetteIndexEntry{CAEN Tutorial}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---


# Introduction
This package provides a feature selection method for single-cell RNA-seq 
data. It encodes the category number and calculate the spearman 
correlation coefficient. 

The encoding the category number (CAEN) method is developed for selecting 
feature genes, which is differentially expressed between classes. 
We have implemented the CAEN method via a set of R functions. 
We make a R package named CAEN, and give a tutorial for the package. 
The method consist three steps.

Step 1: Data Pre-processing;  

Step 2: Encoding the category number;Calculating the spearman 
correlation coefficient between the gene and category number.

Step 3: Calculate the classification error using the genes selected

We employ a simulation dataset to illustrate the usage of the CAEN package.
The programs can run under the Linux system and windows 10 system. The
R versions should larger than 3.6.0.

# Preparations
To install the CAEN package into your R environment, start R and
enter:
```{r, eval=FALSE}
install.packages("BiocManager")
BiocManager::install("CAEN")
```
Then, the CAEN package is ready to load.
```{r}
library(ENTC)
library(SummarizedExperiment)
```

# Data format
In order to reproduce the presented ENTC workflow, the package includes the 
example data sets, which is generated by function *newCountDataSet()*. The 
next we will give an example for how to generate simulation dataset:

```{r}
sim_data <-newCountDataSet(n=100,p=500,K=4,param=10,sdsignal=2,drate=0.2)
```
There are two dataset in the data subdirectory of ENTC package, that is 
**realData** and **sim_data**. **realData** is a real dataset come from *
Post-transcriptional regulation by the let-7 microRNA and the TRIM-NHL
protein LIN41* by Aeschimann F, Gaidatzis D, GroÃŸhans H. 
**sim_data** is a simulation dataset, which is generated 
by *newCountDataSet()* according to the previous steps.

```{r}
data(realData)
realData
```

**realData** includes 44 columns and 20182 rows.That is,there are 44 
samples and 20182 genes in this dataset.

```{r}
data(sim_data)
sim_data
```

**sim_data** includes 100 columns and 481 rows.

# Calculate the spearman correlation coefficient for data
For the category number, we need to consider not only the difference 
between class but also the Intra-category difference. Therefore, we 
propose CAEN method, by encoding the category number within class, 
it get the optimal category number and select the most important genes 
used for classification.
```{r}
data(sim_data)
myscore<-CAEN(assay(sim_data),y=y,K=4,gene_no_list=100)
myscore$np
```


# Calculate classification error rate using genes selected with CAEN method
Getting the important gene, we Calculate classification error rate 
using genes selected. The step is as follows:
```{r}
data(sim_data)
prob<-estimatep(assay(sim_data),y,xte=assay(sim_data),beta=1,
                    type=c("mle","deseq","quantile"), prior=NULL)
prob0<-estimatep(assay(sim_data),y,assay(test_data),beta=1,
                    type=c("mle","deseq","quantile"), prior=NULL)
probb<-t(prob)
probb0<-t(prob0)
train<-assay(sim_data)
test<-assay(test_data)
ddd<-myscore$np
datx<-t(train[ddd,])
datxte<-t(test[ddd,])
zipldacv.out <- ZIPDA.cv(datx,y,prob0=probb)
ZIPLDA.out <- ZIPLDA(datx,y,datxte,transform=FALSE,prob0=probb0,
                rho=zipldacv.out$bestrho)
class<-ZIPLDA.out$ytehat
```
















